### PR TITLE
#3 add Elastic APM

### PR DESCRIPTION
**Pull request by nestarone**

### CODE REVIEW

**Review by stjimmy54 — CHANGES_REQUESTED** [review]

File: ToDoList/Infrastructure/IMediatr.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿namespace ToDoList.Infrastructure.Database
+{
+    internal interface IMediatr
```
[the remark this answers is not in the document]
Comment (nestarone) [reply ToDoList/Infrastructure/IMediatr.cs]: Really, thanks. I deleted unnecessary file.

File: api/obj/.EntityFrameworkCore.targets [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="GetEFProjectMetadata">
```
[the remark this answers is not in the document]
Comment (nestarone) [reply api/obj/.EntityFrameworkCore.targets]: This mistake i fix it. Thanks.

File: ToDoList/Api/Program.cs [hunk 3]
```diff
@@ -19,12 +22,16 @@
     builder.Services.AddScoped(pair.InterfaceType, pair.ValidatorType);
 });
 builder.Services.AddMediatR(Assembly.GetExecutingAssembly())
-    .AddTransient(typeof(IPipelineBehavior<,>), typeof(ValidationBehavior<,>));
+    .AddTransient(typeof(IPipelineBehavior<,>), typeof(ValidationBehavior<,>))
+    .AddTransient(typeof(IPipelineBehavior<,>), typeof(ApmBehavior<,>)); ;
+
```
[the remark this answers is not in the document]
Comment (nestarone) [reply ToDoList/Api/Program.cs:36]: ok, fixed.

File: api/obj/ToDoList.Api.csproj.EntityFrameworkCore.targets [hunk 4]
```diff
@@ -0,0 +1,28 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="GetEFProjectMetadata">
+    <MSBuild Condition=" '$(TargetFramework)' == '' "
```
Comment (stjimmy54) [inline api/obj/ToDoList.Api.csproj.EntityFrameworkCore.targets:4]: Remove this file from git

File: ToDoList/Api/Pipelines/ValidationBehavior.cs [hunk 5]
```diff
@@ -1,39 +1,63 @@
...
+                        {
+                            Key = propertyName,
+                            Values = errorMessages.Distinct().ToArray()
+                        })
+                    .ToDictionary(x => x.Key, x => x.Values);
+                if (errorsDictionary.Any())
+                {
+                    throw new Exception($"Command validation error for type {typeof(TRequest).Name}",
+                        new ValidationException(errorsDictionary.FirstOrDefault().Value.FirstOrDefault()));
+                    //throw new ValidationException(errorsDictionary.FirstOrDefault().Value.ToString());
```
Comment (stjimmy54) [inline ToDoList/Api/Pipelines/ValidationBehavior.cs:48]: Remove this comment

File: ToDoList/Api/Program.cs [hunk 6]
```diff
@@ -45,3 +71,29 @@
...
+        .CreateLogger();
+}
+
+ElasticsearchSinkOptions ConfegureElastikSinks(IConfigurationRoot configuration)
+{
+    return new ElasticsearchSinkOptions(new Uri(configuration["ElasticConfiguration:Uri"]))
+    {
+        AutoRegisterTemplate = true,
+        AutoRegisterTemplateVersion = AutoRegisterTemplateVersion.ESv7,
+        IndexFormat = "logstash-todolist-api-{0:yyyy:MM:dd}"
```
Comment (stjimmy54) [inline ToDoList/Api/Program.cs:97]:
```suggestion
        IndexFormat = "logstash-todolist-api-{0:yyyy.MM.dd}"
```

File: ToDoList/Api/Program.cs [hunk 7]
```diff
@@ -45,3 +71,29 @@
 }
 
 app.Run();
+
```
Comment (stjimmy54) [inline ToDoList/Api/Program.cs:74]: Remove extra blank line

File: ToDoList/Api/Program.cs [hunk 8]
```diff
@@ -1,12 +1,24 @@
...
 using Microsoft.EntityFrameworkCore;
+using Serilog;
+using Serilog.Exceptions;
+using Serilog.Sinks.Elasticsearch;
 using System.Reflection;
 using ToDoList.Api.Pipelines;
 using ToDoList.Infrastructure.Database;
+using Elastic.Apm.SerilogEnricher;
+
+
```
Comment (stjimmy54) [inline ToDoList/Api/Program.cs:15]: Remove extra blank lines

File: ToDoList/Api/appsettings.json [hunk 9]
```diff
@@ -2,11 +2,30 @@
   "ConnectionStrings": {
     "Main": "Host=localhost;Database=todolist;Username=myuser;Password=mypassword"
   },
-  "Logging": {
-    "LogLevel": {
+  //"Logging": {
```
Comment (stjimmy54) [inline ToDoList/Api/appsettings.json:5]: Don't keep comments.

**Review by stjimmy54 — CHANGES_REQUESTED** [review]

File: ToDoList/Infrastructure/DatabaseContext.cs [hunk 10]
```diff
@@ -1,14 +1,85 @@
...
+                .Select(x => x.Entity)
+                .ToList();
+
+            foreach (var entity in entities)
+            {
+                foreach (var @event in entity.DomainEvents!)
+                {
+                    var notificationType = typeof(DomainNotification<,>).MakeGenericType(entity.GetType(), @event.GetType());
+                    var notification = Activator.CreateInstance(notificationType, entity, @event);
+                    await _mediator.Publish(notification!);
```
Comment (stjimmy54) [inline ToDoList/Infrastructure/DatabaseContext.cs:76]:
```suggestion
                    await _mediator.Publish(notification!, cancellationToken);
```

File: ToDoList/Infrastructure/DatabaseContext.cs [hunk 11]
```diff
@@ -1,14 +1,85 @@
...
+                }
+            }
+
```
Comment (stjimmy54) [inline ToDoList/Infrastructure/DatabaseContext.cs:79]: remove blank lines

**Review by stjimmy54 — CHANGES_REQUESTED** [review]

File: ToDoList/Api/Program.cs (hunk 7 above)
Comment (nestarone) [reply ToDoList/Api/Program.cs:71]: Done

File: ToDoList/Infrastructure/DatabaseContext.cs (hunk 11 above)
Comment (nestarone) [reply ToDoList/Infrastructure/DatabaseContext.cs:79]: Done

File: ToDoList/Api/appsettings.json (hunk 9 above)
Comment (nestarone) [reply ToDoList/Api/appsettings.json]: fixed

File: ToDoList/Api/Program.cs (hunk 8 above)
Comment (nestarone) [reply ToDoList/Api/Program.cs]: Done

File: ToDoList/Infrastructure/DatabaseContext.cs (hunk 10 above)
Comment (nestarone) [reply ToDoList/Infrastructure/DatabaseContext.cs]: fixed

File: ToDoList/Api/Features/Todos/AddTodoListCommand.cs [hunk 12]
```diff
@@ -7,17 +7,16 @@ namespace ToDoList.Api.Features.Todos
...
+        public string Title { get; set; }
+        public string Content { get; set; }
     }
 
     public class AddTodoListCommandValidator : AbstractValidator<AddTodoListCommand>
     {
         public AddTodoListCommandValidator()
         {
-            RuleFor(c => c.title).NotEmpty();
-            //RuleFor(c => c.userId).NotNull();
```
Comment (stjimmy54) [inline ToDoList/Api/Features/Todos/AddTodoListCommand.cs:20]: Why are you remove UserId validation? 
```RuleFor(c => c.UserId).NotEmpty();```

File: ToDoList/Api/Features/Todos/DeleteTodoListCommand.cs [hunk 13]
```diff
@@ -7,19 +7,10 @@ namespace ToDoList.Api.Features.Todos
...
     {
-        public Guid userId { get; set; }
-        public Guid todoListItemId { get; set; }
+        public Guid UserId { get; set; }
+        public Guid TodoListItemId { get; set; }
     }
 
-    //public class DeleteTodoListCommandValidator : AbstractValidator<DeleteTodoListCommand>
-    //{
-    //    public DeleteTodoListCommandValidator()
```
Comment (stjimmy54) [inline ToDoList/Api/Features/Todos/DeleteTodoListCommand.cs:16]: Rollback validator

File: ToDoList/Api/Features/Todos/GetTodoListQuery.cs [hunk 14]
```diff
@@ -4,20 +4,16 @@
...
         public Guid Id { get; set; }
         public string Title { get; set; }
         public string Content { get; set; }
-
     }
 
     public class GetTodoListQuery:IRequest<IReadOnlyCollection<TodoListResponse>>
     {
-        public Guid userId { get; set; }
+        public Guid UserId { get; set; }
```
Comment (stjimmy54) [inline ToDoList/Api/Features/Todos/GetTodoListQuery.cs:16]: Add validator

File: ToDoList/Api/Features/Todos/UpdateTodoListCommand.cs [hunk 15]
```diff
@@ -7,18 +7,17 @@ namespace ToDoList.Api.Features.Todos
...
+        public string Title { get; set; }
+        public string Content { get; set; }
     }
 
     public class UpdateTodoListCommandValidator : AbstractValidator<UpdateTodoListCommand>
     {
         public UpdateTodoListCommandValidator()
         {
-            RuleFor(c => c.title).NotEmpty();
-            //RuleFor(c => c.userId).NotNull();
```
Comment (stjimmy54) [inline ToDoList/Api/Features/Todos/UpdateTodoListCommand.cs:21]: ```RuleFor(c => c.UserId).NotEmpty();```

File: ToDoList/Api/Program.cs [hunk 16]
```diff
@@ -45,3 +68,28 @@
...
+        .Enrich.WithEnvironmentName()
+        .Enrich.WithElasticApmCorrelationInfo()
+        .Enrich.WithProperty("ServiceName", "todolist-api")
+        .WriteTo.Console()
+        .WriteTo.Elasticsearch(ConfegureElastikSinks(configuration))
+        .ReadFrom.Configuration(configuration)
+        .CreateLogger();
+}
+
+ElasticsearchSinkOptions ConfegureElastikSinks(IConfigurationRoot configuration)
```
Comment (stjimmy54) [inline ToDoList/Api/Program.cs:87]:
```suggestion
ElasticsearchSinkOptions ConfigureElasticSink(IConfigurationRoot configuration)
```

**Review by stjimmy54 — CHANGES_REQUESTED** [review]

File: ToDoList/Infrastructure/DatabaseContext.cs [hunk 17]
```diff
@@ -1,14 +1,84 @@
...
+                {
+                    var notificationType = typeof(DomainNotification<,>).MakeGenericType(entity.GetType(), @event.GetType());
+                    var notification = Activator.CreateInstance(notificationType, entity, @event);
+                    await _mediator.Publish(notification!, cancellationToken);
+                }
+            }
+
+            return result;
+        }
+
```
Comment (stjimmy54) [inline ToDoList/Infrastructure/DatabaseContext.cs:82]: Remove blank line

File: ToDoList/Api/Features/Todos/AddTodoListCommand.cs (hunk 12 above)
Comment (nestarone) [reply ToDoList/Api/Features/Todos/AddTodoListCommand.cs:20]: fixed

File: ToDoList/Api/Features/Todos/DeleteTodoListCommand.cs (hunk 13 above)
Comment (nestarone) [reply ToDoList/Api/Features/Todos/DeleteTodoListCommand.cs:16]: Done

File: ToDoList/Api/Features/Todos/GetTodoListQuery.cs (hunk 14 above)
Comment (nestarone) [reply ToDoList/Api/Features/Todos/GetTodoListQuery.cs:17]: Done

File: ToDoList/Api/Features/Todos/UpdateTodoListCommand.cs (hunk 15 above)
Comment (nestarone) [reply ToDoList/Api/Features/Todos/UpdateTodoListCommand.cs:21]: Done

File: ToDoList/Api/Program.cs (hunk 16 above)
Comment (nestarone) [reply ToDoList/Api/Program.cs]: Done

File: ToDoList/Infrastructure/DatabaseContext.cs (hunk 17 above)
Comment (nestarone) [reply ToDoList/Infrastructure/DatabaseContext.cs]: Done

File: ToDoList/Api/Features/Todos/DeleteTodoListCommand.cs [hunk 18]
```diff
@@ -7,18 +7,18 @@ namespace ToDoList.Api.Features.Todos
...
-    //    {
-    //        RuleFor(c => c.userId).NotNull();
-    //        RuleFor(c => c.todoListItemId).NotNull();
-    //    }
-    //}
+    public class DeleteTodoListCommandValidator : AbstractValidator<DeleteTodoListCommand>
+    {
+        public DeleteTodoListCommandValidator()
+        {
+            RuleFor(x => x.UserId).NotNull();
```
Comment (stjimmy54) [inline ToDoList/Api/Features/Todos/DeleteTodoListCommand.cs:18]: Guid can't be null. This is struct.
Use NotEmpty() validator

**Review by stjimmy54 — CHANGES_REQUESTED** [review]

File: ToDoList/Api/Features/Todos/DeleteTodoListCommand.cs (hunk 18 above)
Comment (nestarone) [reply ToDoList/Api/Features/Todos/DeleteTodoListCommand.cs]: fixed

File: ToDoList/Infrastructure/DatabaseContext.cs [hunk 19]
```diff
@@ -1,14 +1,82 @@
...
+            modelBuilder.ApplySnakeCaseNamingConvention();
         }
 
         public DbSet<User> Users { get; set; }
+
+        public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken = default)
+        {
+            return await SaveChangesAsync(true, cancellationToken);
+        }
+        public override async Task<int> SaveChangesAsync(bool acceptAllChangesOnSuccess, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (stjimmy54) [inline ToDoList/Infrastructure/DatabaseContext.cs:31]: Add blank line

File: ToDoList/Infrastructure/DatabaseContext.cs [hunk 20]
```diff
@@ -1,14 +1,82 @@
...
+            foreach (var entity in entities)
+            {
+                foreach (var @event in entity.DomainEvents!)
+                {
+                    var notificationType = typeof(DomainNotification<,>).MakeGenericType(entity.GetType(), @event.GetType());
+                    var notification = Activator.CreateInstance(notificationType, entity, @event);
+                    await _mediator.Publish(notification!, cancellationToken);
+                }
+            }
+            return result;
```
Comment (stjimmy54) [inline ToDoList/Infrastructure/DatabaseContext.cs:79]: add blank line before return statement.

File: ToDoList/Api/Program.cs [hunk 21]
```diff
@@ -45,3 +68,28 @@
...
+{
+    Log.Logger = new LoggerConfiguration()
+        .Enrich.FromLogContext()
+        .Enrich.WithExceptionDetails()
+        .Enrich.WithMachineName()
+        .Enrich.WithEnvironmentName()
+        .Enrich.WithElasticApmCorrelationInfo()
+        .Enrich.WithProperty("ServiceName", "todolist-api")
+        .WriteTo.Console()
+        .WriteTo.Elasticsearch(ConfegureElastikSink(configuration))
```
Comment (stjimmy54) [inline ToDoList/Api/Program.cs:82]: `ConfigureElasticSink`

**Pull request merged**